src/DictList.elm

```diff
@@ -937,6 +947,77 @@ fromDict dict =
     DictList dict (Dict.keys dict)
 
 
+{-| Takes a key-fn and a list.
+Creates a `Dict` which maps the key to a list of matching elements.
+    mary = {id=1, name="Mary"}
+    jack = {id=2, name="Jack"}
+    jill = {id=1, name="Jill"}
+    groupBy .id [mary, jack, jill] == DictList.fromList [(1, [mary, jill]), (2, [jack])]
+-}
+groupBy : (comparable1 -> a -> comparable2) -> DictList comparable1 a -> DictList comparable2 (List a)
+groupBy keyfn list =
```

Comment:
Actually, I think what we'd be looking for here is something like this instead:

```elm
groupBy : (a -> comparable) -> List a -> DictList comparable (List a)
```

That is, I think the input should still be a `List`, not another `DictList`. So, the signature would be the same except that we produce a `DictList` instead of a `Dict`. And, the additional behaviour would be that the resulting `DictList` preserves the order of the incoming list (which `List.Extra.groupBy` would not do).

Part of the logic is that we're in `Dict.Extra`, so it is (generally speaking) the `Dict` parameters that we want to be `DictList` -- the `List` things here really are meant to be good-old-lists (though there may be some exceptions).

Or, to put it another way, the point of `groupBy` is to add keys to a list of values, so we need not require that we already have keys.